### DISCUSSION TITLE
The installation hangs

### DISCUSSION THREAD

**Discussion by Yaojian** [prisma/prisma#5646] [Q&A] [answered]
The command `yarn add prisma --dev --verbose` hangs at the last stage for hours

```
verbose 0.2449488 Checking for configuration file "D:\\temp\\pris\\.npmrc".
verbose 0.2462529 Checking for configuration file "C:\\Users\\jyao\\.npmrc".
verbose 0.247281 Found configuration file "C:\\Users\\jyao\\.npmrc".
verbose 0.2489636 Checking for configuration file "C:\\Program Files\\nodejs\\etc\\npmrc".
verbose 0.253584 Checking for configuration file "D:\\temp\\pris\\.npmrc".
verbose 0.2602576 Checking for configuration file "D:\\temp\\.npmrc".
verbose 0.2644845 Checking for configuration file "D:\\temp\\pris\\.yarnrc".
verbose 0.2655629 Checking for configuration file "C:\\Users\\jyao\\.yarnrc".
verbose 0.2666124 Found configuration file "C:\\Users\\jyao\\.yarnrc".
verbose 0.2682038 Checking for configuration file "C:\\Program Files\\nodejs\\etc\\yarnrc".
verbose 0.2693428 Checking for configuration file "D:\\temp\\pris\\.yarnrc".
verbose 0.2702555 Checking for configuration file "D:\\temp\\.yarnrc".
verbose 0.2758719 current time: 2021-02-14T15:17:42.512Z
info No lockfile found.
[1/4] Resolving packages...
verbose 0.3661073 Performing "GET" request to "https://registry.npm.taobao.org/prisma".
verbose 1.7307006 Request "https://registry.npm.taobao.org/prisma" finished with status code 200.
verbose 1.7364046 Performing "GET" request to "https://registry.npm.taobao.org/@prisma%2fengines".
verbose 1.7631568 Request "https://registry.npm.taobao.org/@prisma%2fengines" finished with status code 200.
[2/4] Fetching packages...
[3/4] Linking dependencies...
verbose 1.8026585 Creating directory "D:\\temp\\pris\\node_modules\\@prisma\\engines".
verbose 1.804135 Creating directory "D:\\temp\\pris\\node_modules\\prisma".
verbose 1.8085312 Creating directory "D:\\temp\\pris\\node_modules\\prisma\\build".
.... (omitted)...
verbose 2.0180709 Copying "C:\\Users\\jyao\\AppData\\Local\\Yarn\\Cache\\v6\\npm-prisma-2.16.1-cf73322e5426aabcd084a42dac89866006f5823f-integrity\\node_modules\\prisma\\build\\public\\icons\\tick.svg" to "D:\\temp\\pris\\node_modules\\prisma\\build\\public\\icons\\tick.svg".
[4/4] Building fresh packages...
[1/2] ⠠ @prisma/engines
[-/2] ⠐ waiting...
```

OS:  Windows 10
node: 12.20.2
yarn: 1.22.10

any ideas?

**Comment by timsuchanek** [comment]
Thanks for reporting @Yaojian! Are you using a custom npm registry? From the first glance I can't see the problem yet.

**Reply by Jolg42** [reply]
Could be related to the postinstall hook trying to run generate.

Try to set the env var `SKIP_GENERATE=true` like:

`SKIP_GENERATE=true yarn add prisma --dev --verbose`

**Reply by Jolg42** [reply]
Did you try `SKIP_GENERATE=true`? 

Note: it's for the postinstall in `@prisma/client`

**Comment by Yaojian** [comment, marked as answer]
Does the postinstall download something without using npm's http_proxy? 

I am in China and using a custom npm registry   https://registry.npm.taobao.org/

**Reply by Jolg42** [reply]
So yes there is some download logic in `@prisma/engines` (https://github.com/prisma/engines-wrapper/blob/master/packages/engines/src/download.ts)

The Prisma engines are downloaded from a Prisma owned domain `https://binaries.prisma.sh` during postinstall but the logic does use `http-proxy-agent` / `https-proxy-agent`
https://github.com/prisma/prisma/blob/61c25890411285d01c5d3df07283a79aa7e01cdb/src/packages/fetch-engine/src/downloadZip.ts#L43

**Reply by Yaojian** [reply]
In China, developers are behind the [Great FireWall](https://en.wikipedia.org/wiki/Great_Firewall). It's very slow (and sometimes impossible) to access the original npm repository and may have trouble while accessing many other sharing services.

A quick look in `downloadZip.ts` and [`getProxyAgent.ts`](https://github.com/prisma/prisma/blob/61c25890411285d01c5d3df07283a79aa7e01cdb/src/packages/fetch-engine/src/getProxyAgent.ts) indicates the downloader retrieves the proxy settings from `process.env` but not from npm/yarn.

May be it's better to let the yarn/npm's config to take the precedence? So developers in China can install prisma without changing the system proxy settings.

**Reply by Yaojian** [reply]
By the way, I found the `generate` command also needs to download something that requires a http_proxy set in env :(

It will be easier for me if the prisma CLI provides a `proxy` argument rather than retrieve from env.

**Reply by Jolg42** [reply]
Having a global env should work though no?

**Reply by Yaojian** [reply]
Yes, the global env works.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
